frontend/lib/src/WidgetStateManager.ts

```diff
@@ -282,53 +313,114 @@ export class WidgetStateManager {
+    // reruns receive an "empty" value. With the introduction of batched
+    // trigger handling, we align ChatInput with the same mechanism used by
+    // `setTriggerValue` to avoid race conditions when multiple updates are
+    // emitted within the same macrotask.
+    // ------------------------------------------------------------------
+
+    // 1. Store the value in a temporary WidgetState proto.
     this.createWidgetState(widget, source).chatInputValue = new ChatInputValue(
       value
     )
-    this.onWidgetValueChanged(widget.formId, source, fragmentId)
-    this.deleteWidgetState(widget.id)
+
+    // 2. Mark this widget ID so that it is cleaned-up after the pending
+    //    batch flush. The `scheduleFlush` helper already takes care of
+    //    deleting all IDs present in `pendingTriggerIds` once the update
+    //    message has been sent.
+    this.pendingTriggerIds.add(widget.id)
+
+    // 3. Schedule (or reuse) a macrotask-level flush so that ChatInput
+    //    updates are coalesced with other trigger/value updates that happen
+    //    during the same event loop tick.
+    this.scheduleFlush(fragmentId)
   }
 
   /**
-   * Sets the trigger value for the given widget ID to true, sends a rerunScript message
-   * to the server, and then immediately unsets the trigger value.
+   * 1. Boolean trigger
+   *    setTriggerValue(widgetInfo, { fromUi: true }, fragmentId)
+   *
+   * 2. Payload (JSON-encoded) trigger
+   *    setTriggerValue(widgetInfo, { fromUi: true }, fragmentId, payload)
+   *
+   *    `payload` can be any JSON-serialisable value. For Bidi Component v2
+   *    we always transport payloads via the protobuf `json_trigger_value`
+   *    field as a JSON-stringified array of payload objects. Multiple calls
+   *    within the same macrotask are batched into that array. If `payload`
+   *    is omitted (or `undefined`) we fall back to the boolean
+   *    `trigger_value=true` behaviour.
    */
-  public setTriggerValue(
+
+  public setTriggerValue<T extends Record<string, unknown>>(
     widget: WidgetInfo,
     source: Source,
-    fragmentId: string | undefined
+    fragmentId: string | undefined,
+    value?: T
   ): Promise<void> {
-    this.createWidgetState(widget, source).triggerValue = true
-    return this.setTriggerValueAtEndOfEventLoop(widget, source, fragmentId)
+    // If we already have a pending trigger for this widget in the current
+    // macrotask, append to it instead of overwriting so multiple triggers are
+    // delivered in a single backend message.
+    let widgetState = this.getWidgetState(widget)
+    if (widgetState === undefined) {
+      widgetState = this.createWidgetState(widget, source)
+    }
+
+    if (value === undefined) {
+      // Simple boolean trigger.
+      widgetState.triggerValue = true
+    } else {
+      // Custom Components v2: always encode payloads as a JSON array.
+      // We keep the payload as an object and stringify only once when assigning.
```

Comment:
Are we though? It looks like we are regularly stringifying and parsing below. Probably not a big deal performance wise practically, but should we only be stringifying once we are ready to send the JSON back?